### DISCUSSION TITLE
Full parallel MLP using FINN?

### DISCUSSION THREAD

**Discussion by JCastelli326** [Xilinx/finn#1190] [Q&A] [answered]
Dear FINN community,

I am trying to use finn for a very small MLP that would process 3 or 4 components every clock cycles. Like 20 neurons on the first layer and 8 on the second. I would like to not have the inputs given sequentially to each layer but rather have all layers pipelined and processing in parallel the inputs of the preceding layer. I am struggling to find the right configuration of FINN to reach my goal. Is it even possible? 
We tried multiple PE and SIMD settings without success. Does anyone have some pointers or ideas? 

The goal would be to have a very efficient MLP in terms of throughput (1 sample per clock cycle), and reuse the great work you have been doing. 

Thank you for your support,
Jonathan

**Comment by auphelia** [comment]
Hi @JCastelli326,
Are you able to build a bitfile or a stitched IP for your design?
Were you able to max out PE and SIMD or did you run into limitations there?

If you are dealing with a very small MLP, [LogicNets](https://github.com/Xilinx/logicnets) could also be an interesting project for you.

**Reply by fanfeum** [reply]
Hi @auphelia,
Could you advise on how to max out PE and SIMD for small MLPs?
Should it be done for each layer independently?
I guess it also depends on the number of input features and neurons per each layer, right?

**Reply by auphelia** [reply, marked as answer]
Hi @fanfeum,
We have a great tutorial on manual folding here: https://github.com/Xilinx/finn/blob/dev/notebooks/advanced/3_folding.ipynb
And you can also have a look at the advanced builder settings tutorial, which shows how to pass the folding configuration with a json file: https://github.com/Xilinx/finn/blob/dev/notebooks/advanced/4_advanced_builder_settings.ipynb

**Reply by JCastelli326** [reply]
Hi @auphelia , thank you for your answer! Yes, I tried to max out PE and SIMD, but it seems that FINN does not allow to have multiple FIFOs and neurons in parallel (or maybe I did something wrong). I will check out logicNets, it seems like a great option!
